Account for perspective of the scroller when determining a transformed child's contribution to layout overflow

Ahmad-S792 opened the pull request:
#### 7005f16aba81606966d3c9b89a6cb4f764e64167
<pre>
Account for perspective of the scroller when determining a transformed child&apos;s contribution to layout overflow

<a href="https://bugs.webkit.org/show_bug.cgi?id=285282">https://bugs.webkit.org/show_bug.cgi?id=285282</a>
<a href="https://rdar.apple.com/142573993">rdar://142573993</a>

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/602f4cdf9e0944a68fe91d12ead5b2ece186b8e3">https://chromium.googlesource.com/chromium/src.git/+/602f4cdf9e0944a68fe91d12ead5b2ece186b8e3</a>

This patch fixes the case, where WebKit was failing to account for perspective
of the scroller when determining a transformed child&apos;s contribution to layout
overflow and was not drawing any scrollbar at all.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::layoutOverflowRectForPropagation const):
* LayoutTests/overflow/overflow-transform-perspective.html:
* LayoutTests/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/ios/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/glib/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/visionos/TestExpectations: Skip `overflow` directory

&gt; Progression (No Horizontal Scrollbar - matching Chrome):
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:

&gt; Rebaseline:
* LayoutTests/compositing/tiling/perspective-on-scroller-tile-coverage-expected.txt:
* LayoutTests/platform/ios/compositing/tiling/perspective-on-scroller-tile-coverage-expected.txt:
* LayoutTests/svg/compositing/svg-poster-circle.html: Scrollbar Placement change (no rendering difference)

Canonical link: <a href="https://commits.webkit.org/296652@main">https://commits.webkit.org/296652@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24fffdd5dd851704aa252afd6f4667344c01489c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109188 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28846 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114395 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59463 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82975 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112136 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23500 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98340 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63427 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22891 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16483 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92868 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117511 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36232 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26808 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91994 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36604 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94603 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91800 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23378 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36722 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14466 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36128 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41620 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39148 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37505 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/065506d775208bc24ca858313c2473501c525ff5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47032

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d55276edabf52dabfb5f40ee4f8ff74ea7605fd2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47032

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/24fffdd5dd851704aa252afd6f4667344c01489c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47032

Ahmad-S792 committed:
Account for perspective of the scroller when determining a transformed child's contribution to layout overflow

https://bugs.webkit.org/show_bug.cgi?id=285282
rdar://142573993

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/602f4cdf9e0944a68fe91d12ead5b2ece186b8e3

This patch fixes the case, where WebKit was failing to account for perspective
of the scroller when determining a transformed child's contribution to layout
overflow and was not drawing any scrollbar at all.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::layoutOverflowRectForPropagation const):
* LayoutTests/overflow/overflow-transform-perspective.html:
* LayoutTests/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/ios/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/glib/overflow/overflow-transform-perspective-expected.txt:
* LayoutTests/platform/visionos/TestExpectations: Skip `overflow` directory

> Progression (No Horizontal Scrollbar - matching Chrome):
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/text-perspective-001-expected.txt:

> Rebaseline:
* LayoutTests/compositing/tiling/perspective-on-scroller-tile-coverage-expected.txt:
* LayoutTests/platform/ios/compositing/tiling/perspective-on-scroller-tile-coverage-expected.txt:
* LayoutTests/svg/compositing/svg-poster-circle.html: Scrollbar Placement change (no rendering difference)

Canonical link: https://commits.webkit.org/296652@main
